driver-core/src/main/com/mongodb/internal/connection/OidcAuthenticator.java

```diff
@@ -0,0 +1,731 @@
+        assertFalse(connection.opened());
+        String accessToken = getValidCachedAccessToken();
+        if (accessToken != null) {
+            startAsync().run(c -> {
+                authenticateAsyncUsing(connection, connectionDescription, (bytes) -> prepareTokenAsJwt(accessToken), c);
+            }).onErrorIf(e -> triggersRetry(e), c -> {
+                authLockAsync(connection, connectionDescription, c);
+            }).complete(callback);
+        } else {
+            startAsync().run(c -> {
+                authLockAsync(connection, connectionDescription, c);
+            }).complete(callback);
+        }
+    }
+
+    private static boolean triggersRetry(@Nullable final Throwable t) {
+        if (t instanceof MongoSecurityException) {
+            MongoSecurityException e = (MongoSecurityException) t;
+            Throwable cause = e.getCause();
+            if (cause instanceof MongoCommandException) {
+                MongoCommandException commandCause = (MongoCommandException) cause;
+                return commandCause.getErrorCode() == 18;
+            }
+        }
+        return false;
+    }
+
+    private void authenticateAsyncUsing(final InternalConnection connection,
+            final ConnectionDescription connectionDescription, final Function<byte[], byte[]> evaluateChallengeFunction,
+            final SingleResultCallback<Void> callback) {
+        this.evaluateChallengeFunction = evaluateChallengeFunction;
+        super.authenticateAsync(connection, connectionDescription, callback);
+    }
+
+    private void authenticateUsing(
+            final InternalConnection connection,
+            final ConnectionDescription connectionDescription,
+            final Function<byte[], byte[]> evaluateChallengeFunction) {
+        this.evaluateChallengeFunction = evaluateChallengeFunction;
+        super.authenticate(connection, connectionDescription);
+    }
+
+    private void authLock(final InternalConnection connection, final ConnectionDescription description) {
+        Locks.withLock(getMongoCredentialWithCache().getOidcLock(), () -> {
+            while (true) {
+                try {
+                    authenticateUsing(connection, description, (challenge) -> evaluate(challenge));
+                    break;
+                } catch (MongoSecurityException e) {
+                    boolean shouldRetry = triggersRetry(e) && shouldRetryHandler();
+                    if (!shouldRetry) {
+                        throw e;
+                    }
+                }
+            }
+            return null;
+        });
+    }
+
+    private void authLockAsync(final InternalConnection connection, final ConnectionDescription description,
+            final SingleResultCallback<Void> callback) {
+
+        MongoCredentialWithCache mongoCredentialWithCache = getMongoCredentialWithCache();
+        Locks.withLockAsync(
+                mongoCredentialWithCache.getOidcLock(),
```

Comment:
The sync counterpart does not have the `mongoCredentialWithCache` variable. Also, changed formatting to match the formatting in the sync counterpart

```suggestion
        Locks.withLockAsync(getMongoCredentialWithCache().getOidcLock(),
```